diff --git a/lib/bud/lattices.rb b/lib/bud/lattices.rb
index ceb46cb..0c9c73a 100644
--- a/lib/bud/lattices.rb
+++ b/lib/bud/lattices.rb
@@ -417,6 +417,10 @@ class Bud::LatticeWrapper
     scalar_merge(lhs, rhs)
   end
 
+  def do_unsafe_merge(lhs, rhs)
+    lhs.merge_unsafe(rhs)
+  end
+
   def setup_wiring(input, kind)
     if input.class <= Bud::LatticeWrapper
       add_merge_target
@@ -432,7 +436,8 @@ class Bud::LatticeWrapper
   # Merge "i" into @new_delta
   public
   def insert(i, source)
-    @new_delta = do_merge(current_new_delta, i)
+#    @new_delta = do_merge(current_new_delta, i)
+    @new_delta = do_unsafe_merge(current_new_delta, i)
   end
 
   def <=(i)
@@ -506,10 +511,18 @@ class Bud::LatticeWrapper
     result = false
 
     if @new_delta
-      m = do_merge(current_value, @new_delta)
-      if m != current_value
-        @storage = m
-        result = true
+      if @klass <= Bud::HashSetLattice
+        old_size = current_value.reveal.size
+        @storage = do_unsafe_merge(current_value, @new_delta)
+        if @storage.reveal.size != old_size
+          result = true
+        end
+      else
+        m = do_merge(current_value, @new_delta)
+        if m != current_value
+          @storage = m
+          result = true
+        end
       end
     end
 
@@ -827,6 +840,11 @@ class Bud::HashSetLattice < Bud::Lattice
     wrap_unsafe(@v | i.reveal)
   end
 
+  def merge_unsafe(i)
+    @v.merge(i.reveal)
+    self
+  end
+
   morph :intersect do |i|
     wrap_unsafe(@v & i.reveal)
   end
