diff --git a/lib/bud/lattices.rb b/lib/bud/lattices.rb
index f1cd3e4..fa192de 100644
--- a/lib/bud/lattices.rb
+++ b/lib/bud/lattices.rb
@@ -414,6 +414,29 @@ class Bud::SetLattice < Bud::Lattice
     wrap_unsafe(rv)
   end
 
+  morph :theta do |i, lhs_idx, rhs_idx|
+    rv = []
+    @v.each do |a|
+      # i.probe(rhs_idx, a[lhs_idx]).each {|b|
+      #   rv << [a,b]
+      # }
+      rv += i.probe(rhs_idx, a[lhs_idx]).map {|b| [a,b]}
+    end
+    wrap_unsafe(rv)
+  end
+
+  morph :tc do |i|
+    lhs_idx = 1
+    rhs_idx = 0
+    rv = []
+    @v.each do |a|
+      i.probe(rhs_idx, a[lhs_idx]).each do |b|
+        rv << [a[0], b[1], a[2] + b[2]]
+      end
+    end
+    wrap_unsafe(rv)
+  end
+
   morph :pro do |&blk|
     @v.map(&blk)
   end
@@ -421,6 +444,20 @@ class Bud::SetLattice < Bud::Lattice
   ord_map :size do
     Bud::MaxLattice.new(@v.size)
   end
+
+  def probe(idx, v)
+    @ht ||= build_ht(idx)
+    return @ht[v] || []
+  end
+
+  def build_ht(idx)
+    rv = {}
+    @v.each do |i|
+      rv[i[idx]] ||= []
+      rv[i[idx]] << i
+    end
+    rv
+  end
 end
 
 # A set that admits only non-negative numbers. This allows "sum" to be an
@@ -479,6 +516,32 @@ class Bud::HashSetLattice < Bud::Lattice
     @v.map(&blk)
   end
 
+  morph :tc do |i|
+    lhs_idx = 1
+    rhs_idx = 0
+    rv = Set.new
+    @v.each do |a|
+      i.probe(rhs_idx, a[lhs_idx]).each do |b|
+        rv << [a[0], b[1], a[2] + b[2]]
+      end
+    end
+    wrap_unsafe(rv)
+  end
+
+  def probe(idx, v)
+    @ht ||= build_ht(idx)
+    return @ht[v] || []
+  end
+
+  def build_ht(idx)
+    rv = {}
+    @v.each do |i|
+      rv[i[idx]] ||= []
+      rv[i[idx]] << i
+    end
+    rv
+  end
+
   ord_map :size do
     Bud::MaxLattice.new(@v.size)
   end
